### ISSUE TITLE
Add use secondary roles parameter snowflake provider

### ISSUE THREAD

**Issue by rd-tobias-esser** [Snowflake-Labs/terraform-provider-snowflake#2504]
### Terraform CLI and Provider Versions

terraform {
  required_version = ">= 1.5"
  backend "http" {
  }

  required_providers {
    snowflake = {
      source = "Snowflake-Labs/snowflake"
      version = ">= 0.85.0"
    }
  }
}

### Use Cases or Problem Statement

When granting privileges of views and tables I need to set secondary roles

### Proposal

Solve this as a parameter for the resource snowflake_grant_privileges_to_account_role or as a global option in the snowflake provider resource

### How much impact is this issue causing?

High

**Comment by sfc-gh-jcieslak** [comment]
Hey @rd-tobias-esser 
We have the change ready to merge, but It turned out we cannot merge it right now. We have to refactor the way we're passing the Snowflake connection between Terraform operations. To do it right now, we would have to call `use secondary roles` in hundreds of places instead of one. Unfortunately, this has to wait until we refactor every resource to use already already-initialized client instead of creating one per operation.

**Comment by afeld** [comment]
> refactor every resource to use already already-initialized client instead of creating one per operation

Is there an issue where we can follow that?

**Comment by sfc-gh-jcieslak** [comment]
We did the refactor recently, so this should be unblocked. I'll look at it as soon as possible.

**Comment by sfc-gh-jcieslak** [comment]
This issue is the one to look at for this feature. I looked at it yesterday, but there are some issues with the `use secondary roles all` call through Go Snowflake Driver. We have to postpone this for a while. Most likely until [provider config rework](https://github.com/Snowflake-Labs/terraform-provider-snowflake/blob/main/ROADMAP.md#providers-configuration-rework).

**Comment by arsatiki** [comment]
Hello! I have encountered a situation where this could be useful. Specifically, I'd like to create a user with the `USERADMIN` role and at the same time assign a network policy. This has turned out to be quite difficult, unless I grant the _user_ `ALL` secondary roles as default. Effectively that makes all the roles as powerful as `ACCOUNTADMIN`, which is something I'd like to avoid. So I'd like a way to specify a single secondary role.

Ideally I could specify the secondary role only at the resource definition point but I'd be cool if I could even specify it for a provider.

**Comment by sfc-gh-asawicki** [comment]
Hey @arsatiki. You can specify the role in the provider: https://registry.terraform.io/providers/snowflakedb/snowflake/latest/docs#role-3. So you can have a setup where:
- your user does not allow using secondary roles, but these roles are assigned to the user
- aliased provider A uses role X
- aliased provider B uses role Y
It requires having two provider aliases, but it should work in your use case.

**Comment by arsatiki** [comment]
Hello @sfc-gh-asawicki , that is not what I was asking for. I already have multiple separate provider aliases, one for accountadmin, one for useradmin and so on. I am trying to follow the principle of least privilege as much as I can. This includes not enabling the secondary roles for those users because then every provider alias is effectively an accountadmin.

The problem is that the network policy visibility requires a securityadmin role and I'm creating the users with a useradmin. This problem exists in Snowflake itself, not just the provider. 

What I would like is to have a way to get around that permissions problem without sacrificing the principle of least privilege. With SQL, I could do that. I would like to have a similar capability in the provider too.
